rfcs/0000-sentry-architecture.md

```diff
@@ -0,0 +1,84 @@
+## Summary
+[summary]: #summary
+
+Introduce a new "hidden" and "sentry" mode to the daemon for building
+DoS-resilient network topologies.
+
+## Motivation
+[motivation]: #motivation
+
+Block producer availability is important both for healthy consensus (blocks
+have a limited time to be broadcast before they are no longer useful) and also
+for node operator profits. Block producers (and, to a lesser extent, SNARK
+workers) need to be powerful enough to produce their proofs in a reasonable
+amount of time, and have operational security requirements around protecting
+the private key. If a DoS takes a node offline such that they can't broadcast
+their work, or put the CPUs to use, the operator loses money and consensus is
+weakened. It is also risky to directly expose a machine with sensitive key
+material to untrusted internet traffic. Sensitive nodes can be protected from
+these risks by running them in a new "hidden mode". These hidden nodes will
+communicate via nodes running in "sentry mode".
+
+The recommended way to deploy this would be to have several sentries, possibly
+in different cloud providers/datacenters, configured per hidden node.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+New RPCs:
+
+- `Sentry_forward`: sent from a hidden node to a sentry node with some new work, to be broadcast over the gossip net.
+- `Please_sentry`: sent from a hidden node to its sentries periodically as a sort of "keepalive" to ensure the sentry node continues forwarding messages even if it crashes and forgets about us. These should be sent at least once per slot.
+- `Hidden_forward`: sent from a sentry node to a hidden node on all new gossip.
+
+### Hidden Mode
+
+Nodes in hidden mode will not participate in discovery at all, by never calling
+`begin_advertising`. They will also filter out all connections not from their
+configured sentries. These are "soft" mitigations - hidden nodes, when deployed,
+should not be publicly routable at all, or otherwise have the firewall carefully
+configured to only allow communications with sentries. 
+
+Because hidden nodes are not well connected, we can't rely on the usual
+guarantees of the pubsub implementation to receive block gossip. Thus
+`Hidden_forward`: instead of using `subscribe_encode`, new messages will be
+received from this RPC. 
+
+The sentries are configured with a new `--sentry-address MULTIADDR` daemon flag.
```

Comment:
Is the sentry address a separate multiaddr just for the purposes of communicating with the hidden nodes? Can you make this clear in the RFC if this is the case?

Reply:
It is not, it's just the regular address of the sentry node. I will clarify that.